parse.go

```diff
@@ -0,0 +1,80 @@
+package cli
+
+import (
+	"flag"
+	"strings"
+)
+
+type iterativeParser interface {
+	newFlagSet() (*flag.FlagSet, error)
+	useShortOptionHandling() bool
+}
+
+// To enable short-option handling (e.g., "-it" vs "-i -t") we have to
+// iteratively catch parsing errors.  This way we achieve LR parsing without
+// transforming any arguments. Otherwise, there is no way we can discriminate
+// combined short options from common arguments that should be left untouched.
+func parseIter(ip iterativeParser, args []string) (*flag.FlagSet, error) {
+	for {
+		set, err := ip.newFlagSet()
+		if err != nil {
+			return nil, err
+		}
+
+		err = set.Parse(args)
+		if !ip.useShortOptionHandling() || err == nil {
+			return set, err
+		}
+
+		errStr := err.Error()
+		trimmed := strings.TrimPrefix(errStr, "flag provided but not defined: ")
```

Comment:
This is super hacky. Can't we return a:
```
type FlagNotFoundError error {
    error
    flag string
}
```
and then do a type assertion here, having access to the field member that we need?

Reply:
That error is coming straight from the standard library `flag` package, which does not offer anything in the way of sentinel errors or custom types: https://golang.org/pkg/flag/#FlagSet.Parse

If you dig through the source, you'll eventually find this is what we're dealing with in the underlying `parseOne` function (which I can't figure out how to link to directly): 
```go
f.failf("flag provided but not defined: -%s", name)
```

Since that's just wrapping `fmt.Errorf`, I don't think there's anything we can do better here.